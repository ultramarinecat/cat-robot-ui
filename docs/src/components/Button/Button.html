<!DOCTYPE html>

<html>
<head>
  <title>Button.js</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, target-densitydpi=160dpi, initial-scale=1.0; maximum-scale=1.0; user-scalable=0;">
  <link rel="stylesheet" media="all" href="../../../docco.css" />
</head>
<body>
  <div id="container">
    <div id="background"></div>
    
      <ul id="jump_to">
        <li>
          <a class="large" href="javascript:void(0);">Jump To &hellip;</a>
          <a class="small" href="javascript:void(0);">+</a>
          <div id="jump_wrapper">
          <div id="jump_page_wrapper">
            <div id="jump_page">
              
                
                <a class="source" href="../App/App.html">
                  src/components/App/App.js
                </a>
              
                
                <a class="source" href="../App/ErrorBoundary.html">
                  src/components/App/ErrorBoundary.js
                </a>
              
                
                <a class="source" href="../App/i18n.html">
                  src/components/App/i18n.js
                </a>
              
                
                <a class="source" href="../Auth/Auth.html">
                  src/components/Auth/Auth.js
                </a>
              
                
                <a class="source" href="../Auth/AuthActions.html">
                  src/components/Auth/AuthActions.js
                </a>
              
                
                <a class="source" href="../Auth/AuthReducer.html">
                  src/components/Auth/AuthReducer.js
                </a>
              
                
                <a class="source" href="Button.html">
                  src/components/Button/Button.js
                </a>
              
                
                <a class="source" href="../Cat/Cat.html">
                  src/components/Cat/Cat.js
                </a>
              
                
                <a class="source" href="../Cat/CatActions.html">
                  src/components/Cat/CatActions.js
                </a>
              
                
                <a class="source" href="../Cat/CatReducer.html">
                  src/components/Cat/CatReducer.js
                </a>
              
                
                <a class="source" href="../Cat/fulltilt.html">
                  src/components/Cat/fulltilt.js
                </a>
              
                
                <a class="source" href="../Connection/Connection.html">
                  src/components/Connection/Connection.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionActions.html">
                  src/components/Connection/ConnectionActions.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionReducer.html">
                  src/components/Connection/ConnectionReducer.js
                </a>
              
                
                <a class="source" href="../Connection/messages.html">
                  src/components/Connection/messages.js
                </a>
              
                
                <a class="source" href="../Intro/Intro.html">
                  src/components/Intro/Intro.js
                </a>
              
                
                <a class="source" href="../Intro/IntroActions.html">
                  src/components/Intro/IntroActions.js
                </a>
              
                
                <a class="source" href="../Particles/Particles.html">
                  src/components/Particles/Particles.js
                </a>
              
                
                <a class="source" href="../Pulse/Pulse.html">
                  src/components/Pulse/Pulse.js
                </a>
              
                
                <a class="source" href="../Router/Router.html">
                  src/components/Router/Router.js
                </a>
              
                
                <a class="source" href="../Sound/CatSound.html">
                  src/components/Sound/CatSound.js
                </a>
              
                
                <a class="source" href="../Sound/Mute.html">
                  src/components/Sound/Mute.js
                </a>
              
                
                <a class="source" href="../Sound/Sound.html">
                  src/components/Sound/Sound.js
                </a>
              
                
                <a class="source" href="../Sound/TurningSound.html">
                  src/components/Sound/TurningSound.js
                </a>
              
                
                <a class="source" href="../Spinner/Spinner.html">
                  src/components/Spinner/Spinner.js
                </a>
              
                
                <a class="source" href="../Typing/Typing.html">
                  src/components/Typing/Typing.js
                </a>
              
                
                <a class="source" href="../../configureStore.html">
                  src/configureStore.js
                </a>
              
                
                <a class="source" href="../../index.html">
                  src/index.js
                </a>
              
                
                <a class="source" href="../../serviceWorker.html">
                  src/serviceWorker.js
                </a>
              
                
                <a class="source" href="../../setupTests.html">
                  src/setupTests.js
                </a>
              
            </div>
          </div>
        </li>
      </ul>
    
    <ul class="sections">
        
          <li id="title">
              <div class="annotation">
                  <h1>Button.js</h1>
              </div>
          </li>
        
        
        
        <li id="section-1">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-1">&#182;</a>
              </div>
              <p>Button</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>
<span class="hljs-keyword">import</span> React, { memo, useEffect, useRef } <span class="hljs-keyword">from</span> <span class="hljs-string">'react'</span>;
<span class="hljs-keyword">import</span> PropTypes <span class="hljs-keyword">from</span> <span class="hljs-string">'prop-types'</span>;

<span class="hljs-keyword">import</span> Waves <span class="hljs-keyword">from</span> <span class="hljs-string">'node-waves'</span>;
<span class="hljs-keyword">import</span> <span class="hljs-string">'node-waves/src/scss/waves.scss'</span>;

<span class="hljs-keyword">import</span> styles <span class="hljs-keyword">from</span> <span class="hljs-string">'./Button.module.scss'</span>;

<span class="hljs-keyword">const</span> WAVES_CLASS = <span class="hljs-string">'waves-light'</span>;
<span class="hljs-keyword">const</span> WAVES_DURACTION = <span class="hljs-number">500</span>;

<span class="hljs-keyword">export</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">Button</span>(<span class="hljs-params">{ text, disabled = false, onClick }</span>) </span>{
  <span class="hljs-keyword">const</span> el = useRef(<span class="hljs-literal">null</span>);

  useEffect(<span class="hljs-function"><span class="hljs-params">()</span> =&gt;</span> {
    Waves.attach(<span class="hljs-string">`.<span class="hljs-subst">${styles.main}</span>`</span>, WAVES_CLASS);

    Waves.init({
      <span class="hljs-attr">duration</span>: WAVES_DURACTION
    });
  }, []);

  useEffect(<span class="hljs-function"><span class="hljs-params">()</span> =&gt;</span> {
    <span class="hljs-keyword">if</span> (disabled) {
      el.current.classList.add(styles.disabled);
    } <span class="hljs-keyword">else</span> {
      el.current.classList.remove(styles.disabled);
    }
  }, [disabled]);

  <span class="hljs-keyword">return</span> (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">button</span> <span class="hljs-attr">className</span>=<span class="hljs-string">{styles.main}</span> <span class="hljs-attr">onClick</span>=<span class="hljs-string">{onClick}</span> <span class="hljs-attr">ref</span>=<span class="hljs-string">{el}</span> <span class="hljs-attr">type</span>=<span class="hljs-string">"button"</span>&gt;</span>
      {text}
    <span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span></span>
  );
}

Button.propTypes = {
  <span class="hljs-attr">text</span>: PropTypes.string.isRequired,
  <span class="hljs-attr">disabled</span>: PropTypes.bool,
  <span class="hljs-attr">onClick</span>: PropTypes.func.isRequired
};

<span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> memo(Button);</pre></div></div>
            
        </li>
        
    </ul>
  </div>
</body>
</html>
