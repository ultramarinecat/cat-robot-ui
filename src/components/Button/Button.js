//
// Button
//

import React, { memo, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

import Waves from 'node-waves';
import 'node-waves/src/scss/waves.scss';

import styles from './Button.module.scss';

const WAVES_CLASS = 'waves-light';
const WAVES_DURACTION = 500;

export function Button({ text, disabled = false, onClick }) {
  const el = useRef(null);

  useEffect(() => {
    Waves.attach(`.${styles.main}`, WAVES_CLASS);

    Waves.init({
      duration: WAVES_DURACTION
    });
  }, []);

  useEffect(() => {
    if (disabled) {
      el.current.classList.add(styles.disabled);
    } else {
      el.current.classList.remove(styles.disabled);
    }
  }, [disabled]);

  return (
    <button className={styles.main} onClick={onClick} ref={el} type="button">
      {text}
    </button>
  );
}

Button.propTypes = {
  text: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  onClick: PropTypes.func.isRequired
};

export default memo(Button);
