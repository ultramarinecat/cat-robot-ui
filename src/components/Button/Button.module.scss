@charset 'utf-8';

////
//// Button styles
////

@import 'base.scss';

$button-color: $indigo-400;
$button-color-hover: lighten($button-color, 5%);
$button-color-disabled: darken($button-color, 15%);

$button-font-size: rem(18.56);
$button-font-size-tablet: rem(19.2);

$button-height: rem(46);
$button-height-tablet: rem(50);

$button-border-radius: 2px;

$button-padding: rem(35.2);
$button-letter-spacing: rem(0.576);

$button-transition-delay: 0.5s;
$button-transition-timing-function: ease-out;

.main,
%button {
  background-color: $button-color;
  color: $text-color;

  height: $button-height;
  font-size: $button-font-size;
  text-transform: uppercase;

  @extend %z1;

  border: none;
  border-radius: $button-border-radius;
  display: inline-block;
  outline: 0;
  padding: 0 $button-padding;
  text-decoration: none;
  letter-spacing: $button-letter-spacing;
  transition: $button-transition-delay $button-transition-timing-function;

  cursor: pointer;
  -webkit-tap-highlight-color: transparent;

  &:hover {
    background-color: $button-color-hover;
    @extend %z2;
  }

  &:focus {
    outline: 0;
  }

  @media #{$tablet-plus} {
    height: $button-height-tablet;
    font-size: $button-font-size-tablet;
  }
}

.disabled,
%button-disabled {
  background-color: $button-color-disabled;
  color: $button-text-color-disabled;

  box-shadow: none;
  cursor: default;
  pointer-events: none;

  &:hover {
    background-color: $button-color-disabled;
    color: $button-text-color-disabled;
  }
}
