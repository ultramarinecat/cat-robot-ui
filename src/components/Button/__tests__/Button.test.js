import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import { Button } from '../Button.js';

const DISABLED_CLASS = 'disabled';

describe('Button', () => {
  it('should display a button with text', () => {
    const text = 'cat';
    const { container, getByText } = render(<Button text={text} onClick={jest.fn()} />);

    const button = getByText(text);

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent(text);
    expect(button).not.toHaveClass(DISABLED_CLASS);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be disabled if `disabed` is true', () => {
    const text = 'cat';
    const { container, getByText } = render(
      <Button text={text} disabled={true} onClick={jest.fn()} />
    );

    const button = getByText(text);

    expect(button).toBeInTheDocument();
    expect(button).toHaveClass(DISABLED_CLASS);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call callback prop when clicked', () => {
    const text = 'cat';
    const handleClick = jest.fn();

    const { getByText } = render(<Button text={text} onClick={handleClick} />);
    const button = getByText(text);

    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalled();
  });
});
