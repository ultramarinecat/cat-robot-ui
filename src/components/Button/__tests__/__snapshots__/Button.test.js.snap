// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be disabled if \`disabed\` is true 1`] = `
<button
  class="main waves-effect waves-light disabled"
  type="button"
>
  cat
</button>
`;

exports[`Button should display a button with text 1`] = `
<button
  class="main waves-effect waves-light"
  type="button"
>
  cat
</button>
`;
